### PR TITLE
[alert_handler] Enhancements for mask ROM

### PR DESCRIPTION
**Pull request by msfschaffner**
This adds a couple of alert handler enhancements - specifically:

- Alert enable and assignment to classes is individually lockable
- The alert class configurations are individually lockable
- The alert handler ping timer enable has been separated into its own CSR, instead of tying enablement to a global REGWEN register.

I went through the DIFs and replaced the generated constants that have changed - but I had to disable several DIF tests that now do not work anymore.

However, since the alert enable configuration registers have now all become "non-compact" (i.e. one reg per address), it should be more straightforward to extend the alert handler DIF to support more than 16 alerts (https://github.com/lowRISC/opentitan/issues/3826).

Related: #6041 #6108

Signed-off-by: Michael Schaffner <msf@opentitan.org>

### CODE REVIEW

**Comment by msfschaffner** [comment]
This addresses #6108.

@cindychip and @tjaychen, if this is looking good to you, I will widen the set of reviewers.

**Review by cindychip — COMMENTED** [review]
Thanks Michael, it looks good to me. Thanks for all the DV naming updates!

**Comment by msfschaffner** [comment]
@cfrantz @moidx @tjaychen let me know if this makes sense to you.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [inline hw/ip/alert_handler/data/alert_handler.hjson:252]: we really need a better name for this field...do you have any suggestions?
i know i came up with `regwen_multi`, but i feel like it's very non-intuitive. Is it just me?

Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:252]: `bitwise_regwen`?

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [inline hw/ip/alert_handler/data/alert_handler.hjson:408]: i'm curious, why didn't we make this CLASS0~3 and then use multireg? 
no need to change now, but just trying to remember if there was a specific reason.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [inline hw/ip/alert_handler/data/alert_handler.hjson:491]: this just occurred to me, did we discuss way back then whether the ability to `clear` should be gradually removed?
Ie, we can remove it if we're in phase0, but if we move into phase1, we're going to escalate without stopping? 

It just made me wonder..if we map phase0 to `NMI`, it makes sense to allow clear.  But once we escalate beyond that if no one comes in time, it may not make sense anymore.  

Anyhow, this is just a thought and not really related to this PR.  If it makes sense we can consider it way in the future.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [reply hw/ip/alert_handler/data/alert_handler.hjson:408]: ah i forgot these are templated..so it doesn't even really matter.

File: sw/device/tests/dif/dif_alert_handler_unittest.cc
Comment (tjaychen) [inline sw/device/tests/dif/dif_alert_handler_unittest.cc:143]: @cfrantz can someone help look into this portion? It might be that the `difs` should be written slightly differently with this change also.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (tjaychen) [reply hw/ip/alert_handler/data/alert_handler.hjson:252]: i think that's better... maybe we should take a vote hehe.

**Review by tjaychen — APPROVED** [review]
this LGTM.

File: sw/device/tests/dif/dif_alert_handler_unittest.cc
Comment (moidx) [reply sw/device/tests/dif/dif_alert_handler_unittest.cc:143]: @mcy  is probably the best person to look at the dif changes.

**Review by moidx — COMMENTED** [review]
LG on the interface changes @cfrantz should approve from rom perspective.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:408]: bhu honestly I think this was just the terminology we've used in the spec.

File: hw/ip/alert_handler/data/alert_handler.hjson
Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:491]: No I think the decision whether to allow clearing or not was always pretty black and white.
Either you allow it, or you don't.

We could technically add more knobs here that map to the phases if needed.

Comment (tjaychen) [reply hw/ip/alert_handler/data/alert_handler.hjson:491]: it's something to consider. Because i can imagine during an NMI phase to allow clearing, but disallow if it goes beyond that.
Not critical though.  It's something I would probably file an issue on for future discussions.

Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:491]: It wouldn't be hard to implement. A slight CSR change would work: either make this a 2bit threshold, or provide one bit per phase.

Comment (tjaychen) [reply hw/ip/alert_handler/data/alert_handler.hjson:491]: naw let's not change it yet. I'm not 100% convinced of the utility. It's just something that came to mind and could be good to think about later as the software guys sort through how they want to use it.

Comment (msfschaffner) [reply hw/ip/alert_handler/data/alert_handler.hjson:491]: sounds good!

**Comment by msfschaffner** [comment]
Rebased to make the changes atomic.

File: sw/device/tests/dif/dif_alert_handler_unittest.cc
Comment (cfrantz) [inline sw/device/tests/dif/dif_alert_handler_unittest.cc:818]: ALERT_HANDLER_CLASSB_ **CLR** _REGWEN_REG_OFFSET

And similar for L824 and L832.

**Review by cfrantz — APPROVED** [review]
I spoke to msf@ briefly about this change:
- It sounds like separating ping enables and ping locking is rather difficult and likely to result in escalation glitches.  This means the Mask ROM has to to turn on only the alerts it cares about and later (owner) code cannot turn those alerts off.
- Currently, it seems that alert classes also get locked into their classifications.  If I have to lock alert-enables, I'd like to give later code the ability to re-classify so owners can make their own decisions about the importance of the different alerts.
- Overall, I think the register layout improves the register layout and allows for easier expansion of alerts.

**Comment by tjaychen** [comment]
> I spoke to msf@ briefly about this change:
> 
> * It sounds like separating ping enables and ping locking is rather difficult and likely to result in escalation glitches.  This means the Mask ROM has to to turn on only the alerts it cares about and later (owner) code cannot turn those alerts off.

@cfrantz i thought this was the ROM intent overall? Let me know if I misunderstood. 

> * Currently, it seems that alert classes also get locked into their classifications.  If I have to lock alert-enables, I'd like to give later code the ability to re-classify so owners can make their own decisions about the importance of the different alerts.

similar the above, i think if we allow the re-classification to be done (ie from something fatal to trivial) that could become a problem as well.  Let me know, I thought this was the intent as well. 

> * Overall, I think the register layout improves the register layout and allows for easier expansion of alerts.

File: sw/device/tests/dif/dif_alert_handler_unittest.cc
Comment (msfschaffner) [reply sw/device/tests/dif/dif_alert_handler_unittest.cc:818]: Thanks! I have re-enabled those tests again, since they work now with this change.

**Comment by msfschaffner** [comment]
@tjaychen @cindychip can you merge this one for me once CI completes?

**Comment by tjaychen** [comment]
> @tjaychen @cindychip can you merge this one for me once CI completes?

will do!

File: sw/device/tests/dif/dif_alert_handler_unittest.cc
Comment (timothytrippel) [reply sw/device/tests/dif/dif_alert_handler_unittest.cc:143]: I will look into fixing these as apart of fixing the DIF setting of the enable/class multireg now.

**Pull request merged**
